truffle/src/main/java/org/jruby/truffle/nodes/ext/BigDecimalNodes.java

```diff
@@ -1379,4 +1645,92 @@ public int toISpecial(RubyBasicObject value) {
         }
 
     }
+
+    /**
+     * Casts a value into a BigDecimal.
+     */
+    @NodeChild(value = "value", type = RubyNode.class)
+    @ImportStatic(BigDecimalCoreMethodNode.class)
+    public abstract static class BigDecimalCastNode extends RubyNode {
+        public BigDecimalCastNode(RubyContext context, SourceSection sourceSection) {
+            super(context, sourceSection);
+        }
+
+        public abstract BigDecimal executeBigDecimal(VirtualFrame frame, Object value);
+
+        public abstract Object executeObject(VirtualFrame frame, Object value);
+
+        @Specialization
+        public BigDecimal doInt(long value) {
+            return BigDecimal.valueOf(value);
+        }
+
+        @Specialization
+        public BigDecimal doDouble(double value) {
+            return BigDecimal.valueOf(value);
+        }
+
+        @Specialization(guards = "isRubyBignum(value)")
+        public BigDecimal doBignum(RubyBasicObject value) {
+            return new BigDecimal(BignumNodes.getBigIntegerValue(value));
+        }
+
+        @Specialization(guards = {"isRubyBigDecimal(value)", "isNormal(value)"})
+        public BigDecimal doBigDecimal(RubyBasicObject value) {
+            return (BigDecimal) VALUE_PROPERTY.get(value.getDynamicObject(), true);
+        }
+
+        @Fallback
+        public RubyBasicObject doBigDecimalFallback(Object value) {
+            return nil();
+        }
+        // TODO (pitr 22-Jun-2015): How to better communicate failure without throwing
```

Comment:
This relates to `executeWith`, I can then use the nil value in `BigDecimalCoerceNode`:
```
@Specialization(guards = {"isRubyBigDecimal(value)", "isNil(cast)"})
public RubyBasicObject doBigDecimal(RubyBasicObject value, RubyBasicObject cast) {
    return value;
}
```